File: app/schemas/forecast.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ForecastRunCreateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    forecast_run_id: UUID
    status: str
    scenario_version_id: UUID
    engine_version: str


class ForecastRunOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: UUID
    scenario_version_id: UUID
    engine_version: str
    status: str
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    outputs: Optional[Dict[str, Any]] = None


class ForecastCompareResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scenario_id: UUID
    a_version: int
    b_version: int
    bucket_dates: List[str]
    delta_enrolled_per_bucket: List[float]
    a_cumulative_enrolled: List[float]
    b_cumulative_enrolled: List[float]
    delta_demand: Dict[str, List[float]]
    engine_version: str

File: frontend/src/types/forecast.ts
import type { UUID } from './common';

export interface ForecastRun {
  id: UUID;
  scenario_version_id: UUID;
  engine_version: string;
  status: string;
  started_at: string | null;
  finished_at: string | null;
  outputs: ForecastOutputs | null;
}

export interface ForecastOutputs {
  bucket_dates: string[];
  enrolled_per_bucket: number[];
  cumulative_enrolled: number[];
  demand_per_bucket: Record<string, number[]>;
  visits_per_bucket?: number[];
  [key: string]: unknown;
}

export interface ForecastRunCreateResponse {
  forecast_run_id: UUID;
  status: string;
  scenario_version_id: UUID;
  engine_version: string;
}

export interface ForecastCompare {
  scenario_id: UUID;
  a_version: number;
  b_version: number;
  bucket_dates: string[];
  delta_enrolled_per_bucket: number[];
  a_cumulative_enrolled: number[];
  b_cumulative_enrolled: number[];
  delta_demand: Record<string, number[]>;
  engine_version: string;
}
